firedrake/pointwise_operators.py

```diff
@@ -0,0 +1,813 @@
+        aa = function_space
+        ofs = self.original_function_space()
+        if not isinstance(ofs, functionspaceimpl.WithGeometry):
+            fs1 = functionspaceimpl.FunctionSpace(function_space.mesh().topology, ofs.ufl_element())
+            fs = functionspaceimpl.WithGeometry(fs1, function_space.mesh())
+        else:
+            fs = ofs
+
+        Function.__init__(self, fs, val, name, dtype, count=self._count)  # count has been initialised in ExternalOperator.__init__
+
+        self._ufl_function_space = aa
+        self._original_function_space = fs
+        self.operator_data = operator_data
+
+    @abstractmethod
+    def compute_derivatives(self):
+        """apply the derivatives on operator_data"""
+
+    @abstractmethod
+    def evaluate(self):
+        """define the evaluation method for the ExternalOperator object"""
+
+    def copy(self, deepcopy=False):
+        r"""Return a copy of this CoordinatelessFunction.
+
+        :kwarg deepcopy: If ``True``, the new
+            :class:`CoordinatelessFunction` will allocate new space
+            and copy values.  If ``False``, the default, then the new
+            :class:`CoordinatelessFunction` will share the dof values.
+        """
+        if deepcopy:
+            val = type(self.dat)(self.dat)
+        else:
+            val = self.dat
+        return type(self)(*self.ufl_operands, function_space=self.function_space(), val=val,
+                          name=self.name(), dtype=self.dat.dtype,
+                          derivatives=self.derivatives,
+                          operator_data=self.operator_data)
+
+    # Computing the action of the adjoint derivative may require different procedures
+    # depending on wrt what is taken the derivative
+    # E.g. the neural network case: where the adjoint derivative computation leads us
+    # to compute the gradient wrt the inputs of the network or the weights.
+    def adjoint_action(self, x, idx):
+        r"""Starting from the residual form: F(N(u, m), u(m), m) = 0
+            This method computes the action of (dN/dq)^{*}
+            where q \in \{u, m\}.
+        """
+        derivatives = tuple(dj + int(idx == j) for j, dj in enumerate(self.derivatives))
+        dNdq = self._ufl_expr_reconstruct_(*self.ufl_operands, derivatives=derivatives)
+        dNdq = dNdq.evaluate()
+        dNdq_adj = transpose(dNdq)
+        result = firedrake.assemble(dNdq_adj)
+        return result.vector() * x
+
+    def _adjoint(self, idx):
+        derivatives = tuple(dj + int(idx == j) for j, dj in enumerate(self.derivatives))
+        dNdq = self._ufl_expr_reconstruct_(*self.ufl_operands, derivatives=derivatives)
+        dNdq = dNdq.evaluate()
+        dNdq_adj = transpose(dNdq)
+        result = firedrake.assemble(dNdq_adj)
+        return result
+
+    @utils.cached_property
+    def _split(self):
+        return tuple(Function(V, val) for (V, val) in zip(self.function_space(), self.topological.split()))
```

Comment:
forgets the operator-ness, is this right?

Reply:
It's been a long time but I think we don't want `_split` to return subcomponents as being other pointwise operators themselves. `AbstractPointwiseOperator` is directly inheriting from `Function` so I think this is right.